cmd/noms/diff/diff.go

```diff
@@ -184,102 +205,66 @@ func diffOrdered(w diffWriter, p types.Path, lf lineFunc, ppf pathPartFunc, df d
+			stop = !d.sendDiff(dif)
 		case types.DiffChangeRemoved:
-			writeHeader(w, p, &wroteHdr)
-			err = lf(w, DEL, k, v1(change.V))
+			dif := Difference{Path: p1, ChangeType: types.DiffChangeRemoved, OldValue: v1(change.V), NewValue: nil}
+			stop = !d.sendDiff(dif)
 		case types.DiffChangeModified:
 			c1, c2 := v1(change.V), v2(change.V)
 			if shouldDescend(c1, c2) {
-				writeFooter(w, &wroteHdr)
-				err = diff(w, append(p, ppf(k)), change.V, c1, c2)
+				stop = !d.diff(p1, c1, c2)
 			} else {
-				writeHeader(w, p, &wroteHdr)
-				lf(w, DEL, k, c1)
-				err = lf(w, ADD, k, c2)
+				dif := Difference{Path: p1, ChangeType: types.DiffChangeModified, OldValue: c1, NewValue: c2}
+				stop = !d.sendDiff(dif)
 			}
 		default:
 			panic("unknown change type")
 		}
 	}
 
-	writeFooter(w, &wroteHdr)
-
-	if err != nil {
+	if stop {
 		stopChan <- struct{}{}
-		// Wait for diff to stop.
 		for range changeChan {
 		}
 	}
-	return
-}
 
-func writeHeader(w io.Writer, p types.Path, wroteHdr *bool) error {
-	if *wroteHdr {
-		return nil
-	}
-	*wroteHdr = true
-	hdr := "(root)"
-	if len(p) > 0 {
-		hdr = p.String()
-	}
-	return write(w, []byte(hdr+" {\n"))
+	return
 }
 
-func writeFooter(w io.Writer, wroteHdr *bool) error {
-	if !*wroteHdr {
+func copyPath(p types.Path) types.Path {
+	if len(p) == 0 {
 		return nil
 	}
-	*wroteHdr = false
-	return write(w, []byte("  }\n"))
-}
-
-func line(w io.Writer, op prefixOp, key, val types.Value) error {
-	genPrefix := func(w *writers.PrefixWriter) []byte {
-		return []byte(op)
-	}
-	pw := &writers.PrefixWriter{Dest: w, PrefixFunc: genPrefix, NeedsPrefix: true}
-	if key != nil {
-		writeEncodedValue(pw, key)
-		write(w, []byte(": "))
-	}
-	writeEncodedValue(pw, val)
-	return write(w, []byte("\n"))
+	p1 := make(types.Path, len(p))
+	copy(p1, p)
+	return p1
 }
 
-func field(w io.Writer, op prefixOp, name, val types.Value) error {
-	genPrefix := func(w *writers.PrefixWriter) []byte {
-		return []byte(op)
-	}
-	pw := &writers.PrefixWriter{Dest: w, PrefixFunc: genPrefix, NeedsPrefix: true}
-	write(pw, []byte(name.(types.String)))
-	write(w, []byte(": "))
-	writeEncodedValue(pw, val)
-	return write(w, []byte("\n"))
+// shouldDescend returns true, if Value is not primitive or is a Ref.
+func shouldDescend(v1, v2 types.Value) bool {
+	kind := v1.Type().Kind()
+	return !types.IsPrimitiveKind(kind) && kind == v2.Type().Kind() && kind != types.RefKind
 }
 
-func writeEncodedValue(w io.Writer, v types.Value) error {
-	if v.Type().Kind() != types.BlobKind {
-		return types.WriteEncodedValue(w, v)
+// stopSent returns true if a message has been sent to this StopChannel
+func (d differ) sendDiff(dif Difference) bool {
+	select {
+	case <-d.stopChan:
+		return false
+	default:
 	}
-	write(w, []byte("Blob ("))
-	write(w, []byte(humanize.Bytes(v.(types.Blob).Len())))
-	return write(w, []byte(")"))
-}
 
-func write(w io.Writer, b []byte) error {
-	_, err := w.Write(b)
-	return err
+	d.diffChan <- dif
```

Comment:
This isn't quite right.

Consider if the consumer has not sent a message to the stop chan when this code runs. It falls through the select here, and this code blocks trying to send on the diffChan. *Then* the consumer sends on the stop chan. This will deadlock.

Instead you need to do:
```go
select {
case <-d.stopChan:
  return false
case d.diffChan <- dif:
  return true
}
```